tool/planet/dns.go

```diff
@@ -110,16 +107,48 @@ func updateEnvDNSAddresses(client *kubernetes.Clientset, role agent.Role) error
 	return trace.Wrap(writeEnvDNSAddresses([]string{svcWorker.Spec.ClusterIP, svcMaster.Spec.ClusterIP}, true))
 }
 
-// createService creates the kubernetes DNS service if it doesn't already exist.
-// The service object is managed by gravity, but we create a placeholder here, so that we can read the IP address
-// of the service, and configure kubelet with the correct DNS addresses before starting
-func createService(name string) error {
+func createDNSServices(ctx context.Context, serviceCIDR net.IPNet) error {
 	client, err := cmd.GetKubeClientFromPath(constants.SchedulerConfigPath)
 	if err != nil {
 		return trace.Wrap(err)
 	}
+	ipalloc := ipallocator.NewAllocatorCIDRRange(&serviceCIDR)
+	services := client.CoreV1().Services(metav1.NamespaceSystem)
+	for _, name := range []string{"kube-dns", "kube-dns-worker"} {
+		if err := createDNSService(ctx, name, services, ipalloc); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
 
-	service := &v1.Service{
+// createDNSService creates the kubernetes DNS service if it doesn't already exist.
+// The service object is managed by gravity, but we create a placeholder here, so that we can read the IP address
+// of the service, and configure kubelet with the correct DNS addresses before starting
+func createDNSService(ctx context.Context, name string, services corev1.ServiceInterface, ipalloc *ipallocator.Range) error {
+	ip, err := ipalloc.AllocateNext()
+	if err != nil && err != ipallocator.ErrFull {
```

Comment:
What does ErrFull mean in this case? And why is it not a terminal error here but a terminal error below (inside isIPAlreadyAllocatedError case)?

Reply:
Not really following but it did not need to be `PermanentError` here so I changed it to `trace.Wrap` instead.